vembrane/__init__.py

```diff
@@ -49,52 +49,91 @@
+        self._sample_name = sample_name
         self._sample = sample
-        self._data = format_data
+        self._data = {}
 
     def __getitem__(self, item):
         try:
             return self._data[item]
-        except KeyError as ke:
-            raise UnknownFormatField(self._record_idx, self._sample, ke)
+        except KeyError:
+            try:
+                sample_format = self._sample[item]
+            except KeyError as ke:
+                raise UnknownFormatField(self._record_idx, self._sample_name, ke)
+            self._data[item] = sample_format
+            return sample_format
 
 
 class Format:
-    def __init__(self, record_idx: int, sample_formats: Dict[Sample, Dict[str, Any]]):
+    def __init__(self, record_idx: int, record_samples: List[VariantRecordSample]):
         self._record_idx = record_idx
-        self._sample_formats = sample_formats
+        self._record_samples = record_samples
+        self._sample_formats = {}
 
     def __getitem__(self, item):
         try:
             return self._sample_formats[item]
-        except KeyError as ke:
-            raise UnknownSample(self._record_idx, ke)
+        except KeyError:
+            try:
+                record_sample = self._record_samples[item]
+            except KeyError as ke:
+                raise UnknownSample(self._record_idx, ke)
+            sample = Sample(self._record_idx, item, record_sample)
+            self._sample_formats[item] = sample
+            return sample
 
 
 class Info:
-    def __init__(self, record_idx: int, info_dict: Dict[str, Dict[str, Any]]):
+    def __init__(self, record_idx: int, record_info: VariantRecordInfo, ann_key: str):
         self._record_idx = record_idx
-        self._info_dict = info_dict
+        self._record_info = record_info
+        self._ann_key = ann_key
+        self._info_dict = {}
 
     def __getitem__(self, item):
         try:
             return self._info_dict[item]
-        except KeyError as ke:
-            raise UnknownInfoField(self._record_idx, ke)
+        except KeyError:
+            try:
+                if item == self._ann_key:
+                    raise KeyError(item)
+                untyped_value = self._record_info[item]
+            except KeyError as ke:
+                raise UnknownInfoField(self._record_idx, ke)
+            value = self._info_dict[item] = type_info(untyped_value)
+            return value
 
 
 class Annotation:
-    def __init__(self, record_idx: int, annotation_data: Dict[str, Any]):
+    def __init__(self, ann_key: str, header: VariantHeader):
+        self._record_idx = -1
+        self._annotation_data = {}
+        self._data = {}
+        annotation_keys = get_annotation_keys(header, ann_key)
+        self._ann_conv = {
+            entry.name: (ann_idx, entry.convert)
+            for ann_idx, entry in enumerate(map(ANN_TYPER.get_entry, annotation_keys))
+        }
+
+    def update(self, record_idx: int, annotation: str):
```

Comment:
This masks `Dict.update`. Is that intentional or should we rename that to something like `update_from` or `update_from_record` or `gather_info` or ...?

Reply:
Hm, yes we can rename it. Albeit, `dict.update` should never be called :). TBH, it is a little hacky to have `Environment` derive from `dict`. But it has to be `dict` to be able to act as `globals`.
If we find this too conflated going forward, we can refactor things a bit later, maybe.

Reply:
Well keep it as is for now.

Reply:
Agreed.